prov/efa/docs/efa_rdm_protocol_v4.md

```diff
@@ -0,0 +1,1318 @@
+| 0x40 | `req_opt_connid_hdr`   | This REQ packet has the optional connid header |
+
+**REQ optional headers** contain additional information needed by the receiver of the REQ packets.
+As mentioned earlier, the existence of optional header in a REQ packet is indicated by bits in the `flags`
+field of the base header. There are currently 3 REQ optional headers defined:
+
+1. the raw address header, which has the following format:
+
+Table: 3.2 format of REQ optional raw address header
+
+| Field | type    | Length | C type |
+|-|-|-|-|
+| `size`  | integer | 4      | `uint32` |
+| `addr`  | array   | `size` | `uint8[]` |
+
+As can be seen, the optional raw address is consisted of two fields `size` and `addr`. The field `size` describes
+number of bytes in the `addr` array. The field `addr` contains the raw address. The `size` field is necessary because
+the raw address format of EFA can be expanded in the future.
+
+As mentioned before, an endpoint will include raw address in REQ packet before it receives a handshake packet back
+from a peer. This is because the peer might not have the endpoint's raw address in its address vector, thus cannot
+communicate with the endpoint.
+
+2. the CQ data header, which is an 8 byte integer. CQ data header is used when application called libfabric's
+CQ data send/write API (such as `fi_senddata`, `fi_tsenddata` and `fi_writedata`), which will include an extra
+data in the RX completion entry written to application.
+
+3. the connid (connection ID) header, which is a 4 byte integer. It is used when peer has the "connid header"
+extra request, and the endpoint can support it. More information about this header in section 4.4.
+
+Note, it is possible to have multiple optional REQ headers in one REQ packets. In this case, the order they appear
+in the REQ packets must be the same as their bit appear in the `flags` field. e.g. the raw address header
+must precede the CQ data header, and the CQ data header must precede the connid header.
+
+**Application data** follows immediately after the optional header. Note that not all REQ packet types contain
+application data. For example, the RTR (Request To Read) packet type does not contain application data.
+
+### 3.2 baseline features for two-sided communication
+
+This section describes the 3 baseline features for two sided communication: eager message, medium message, long-CTS message.
+Each of them correspond to the same named sub-protocol. When describing a sub-protocol, we always follow
+the same structure: workflow, packet format and implementation tips.
+
+#### Eager message feature/sub-protocol
+
+Eager message feature/sub-protocol is used when application's send buffer is small enough to be fit in one packet.
+This protocol works in the following order:
+
+1. On sender side, application call libfabric's send API, providing a send buffer.
+2. On receiver side, application call libfabric's receive API, providing a receive buffer.
```

Comment:
Ideally 1 and 2 are flipped here since this order could lead to an unexpected message